internals/daemon/daemon.go

```diff
@@ -194,7 +202,31 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if rspe := access.CheckAccess(c.d, r, ucred, user); rspe != nil {
+	// Optimisation: avoid calling userFromRequest, which acquires the state
+	// lock, in case we don't need to (when endpoint is OpenAccess). This
+	// avoids holding the state lock for /v1/health in particular, which is
+	// not good: https://github.com/canonical/pebble/pull/369
+	var user *UserState
+	if _, isOpen := access.(OpenAccess); !isOpen {
```

Comment:
I'm OK with this special case because it doesn't actually bypass the access checker. As long as we still actually call the access checker, we can have this optimisation.

Reply:
This should be fine. In case there's another implementation in the future of `access` that is not of concrete type `OpenAccess`, but behaves the same, we will only "lose" the optimization (meaning we unnecessarily call `userFromRequest` and populate `user` properly), but it will still do the right thing.